lib/rails_pg_adapter/patch.rb

```diff
@@ -70,37 +76,38 @@ def exec_no_cache(*args)
              ::ActiveRecord::ConnectionNotEstablished,
              ::ActiveRecord::NoDatabaseError => e
         raise unless ::RailsPgAdapter::Patch.supported_errors?(e)
+        handle_schema_cache_error(e)
+        handle_activerecord_error(e)
+        raise unless try_reconnect?(e)
 
-        handle_error(e) unless try_reconnect?(e)
         sleep_time = sleep_times.shift
-        handle_error(e) unless sleep_time
-        warn("Retry query failed, retrying again in #{sleep_time} sec. Retrying: #{query}")
-        sleep(sleep_time)
-        connect
-        retry
+        if sleep_time
+          warn("Retry query failed, retrying again in #{sleep_time} sec. Retrying: #{query}")
+          sleep(sleep_time)
+          connect
+          retry
+        else
+          handle_activerecord_error(e)
+          raise
+        end
       end
     end
 
     def try_reconnect?(e)
       return false if in_transaction?
       return false unless ::RailsPgAdapter::Patch.failover_error?(e.message)
       return false unless ::RailsPgAdapter.reconnect_with_backoff?
-
-      begin
-        disconnect_conn!
-        true
-      rescue ::ActiveRecord::ConnectionNotEstablished
-        false
-      end
+      true
```

Comment:
I like the move towards [CQRS](https://martinfowler.com/bliki/CQRS.html) here 👌

Reply:
🥳